### DISCUSSION TITLE
Using keymapper to map special characters

### DISCUSSION THREAD

**Discussion by MrMufflon** [houmain/keymapper#151] [Q&A] [answered]
Hi,
I want to use keymapper to map special german characters to hotkeys. I'm struggling just to get keymapper to work at all.
I'm on Ubuntu 24.04 on X11. I tried keymapper 4.4.0 at first, but it lead to no keystrokes being forwarded at all. I had to kill the process. I then switched to 4.3.1 and now all keystrokes are forwarded, but my config is ignored. With 4.4.0 and the `-v` flag I saw the correct file being loaded and also the contexts where logged. With 4.3.1 only the config file is logged.
I just tried this as a config `Control{O} >> "ö"` in `~/.config/keymapper.conf`.
Thanks,
Felix

**Comment by houmain** [comment]
Hi, strange that these versions behave differently. I will try them out in a VM soon.
Can you please ensure that only one `keymapper` and one `keymapperd` instance run at the same time and kill them before trying another version. Maybe they were mixed up. The new versioin should make this problem less likely.
Can your keyboard layout type an 'ö'? Did you also try some other key like `Control{O} >> X`? Strings with keys requiring a dead/compose keys are not implemented on Linux yet. 
It should work to simulate the composition like (where Insert is the set [Compose key](https://help.ubuntu.com/community/ComposeKey)):
```
Compose = Insert
Control{O} >> !Any Compose Shift{Quote} O
```

**Reply by ristomatti** [reply]
If layout "English (international with AltGr dead keys)" is an option, this should work:

```ini
Control{A}  >> !Any AltRight{Q}
Control{O}  >> !Any AltRight{P}
```

**Comment by MrMufflon** [comment]
I can confirm, only one instance is running for `keymapper` and `keymapperd`. My keyboard is an english layout, so it can't type the german special characters. I tried normal characters `Control{O} >> "U"` with the same result. I switched back to 4.4.0 with the same behaviour, no keystrokes forwarded.
Thanks for looking into this!

**Reply by houmain** [reply]
I tried the Ubuntu 24.04 live version with X11 and US keyboard layout now but unfortunately could not reproduce a problem, neither with 4.4.0 nor with 4.3.1. With the compose key set to `Insert` and this configuration I could enter an "ö":
```
Compose = Insert
Control{O} >> !Any Compose Shift{Quote} O
```
Can you please run "sudo keymapperd -v" in one terminal and "keymapper -v" in another and post the output?

**Reply by MrMufflon** [reply]
When I do `sudo keymapperd -v` the keyboard inputs stop working, I can't start keymapper after stat. The logs say

Waiting for keymapper to connect
Resetting configuration
Active contexts received (1)
Updating device list
  /dev/input/event0 grabbed (Power Button)
  /dev/input/event1 grabbed (Power Button)
  /dev/input/event7 grabbed (Dygma Raise Consumer Control)
  /dev/input/event9 grabbing failed (Dygma Raise Keyboard)
  /dev/input/event11 ignored (Dygma Raise Mouse)
  /dev/input/event12 ignored (Dygma Raise)
  /dev/input/event13 ignored (Corsair CORSAIR DARK CORE RGB SE Wireless USB Receiver Mouse)
  /dev/input/event14 grabbed (Corsair CORSAIR DARK CORE RGB SE Wireless USB Receiver Consumer Control)
  /dev/input/event16 ignored (Corsair CORSAIR DARK CORE RGB SE Wireless USB Receiver)
  /dev/input/event17 ignored (Corsair CORSAIR DARK CORE RGB SE Wireless USB Receiver)
  /dev/input/event18 ignored (Corsair CORSAIR DARK CORE RGB SE Wireless USB Receiver)
  /dev/input/event19 grabbing failed (Corsair CORSAIR DARK CORE RGB SE Wireless USB Receiver Keyboard)
  /dev/input/event2 grabbing failed (Logitech MX Ergo)
  /dev/input/event3 ignored (Razer Razer DeathAdder V2)
  /dev/input/event4 grabbing failed (Razer Razer DeathAdder V2 Keyboard)
  /dev/input/event6 ignored (Razer Razer DeathAdder V2)
  /dev/input/event20 grabbing failed (Razer Razer DeathAdder V2)
  /dev/input/event21 ignored (HDA NVidia HDMI/DP,pcm=3)
  /dev/input/event22 ignored (HDA NVidia HDMI/DP,pcm=7)
  /dev/input/event23 ignored (HDA NVidia HDMI/DP,pcm=8)
  /dev/input/event24 ignored (HDA NVidia HDMI/DP,pcm=9)
  /dev/input/event25 ignored (HD-Audio Generic Front Mic)
  /dev/input/event26 ignored (HD-Audio Generic Rear Mic)
  /dev/input/event27 ignored (HD-Audio Generic Line)
  /dev/input/event28 ignored (HD-Audio Generic Line Out)
  /dev/input/event29 ignored (HD-Audio Generic Front Headphone)
  /dev/input/event30 grabbed (keyd virtual keyboard)
  /dev/input/event31 ignored (keyd virtual pointer)
Creating virtual device 'Keymapper'
Entering update loop
Updating device list
  /dev/input/event0 already grabbed (Power Button)
  /dev/input/event1 already grabbed (Power Button)
  /dev/input/event7 already grabbed (Dygma Raise Consumer Control)
  /dev/input/event9 grabbing failed (Dygma Raise Keyboard)
  /dev/input/event11 ignored (Dygma Raise Mouse)
  /dev/input/event12 ignored (Dygma Raise)
  /dev/input/event13 ignored (Corsair CORSAIR DARK CORE RGB SE Wireless USB Receiver Mouse)
  /dev/input/event14 already grabbed (Corsair CORSAIR DARK CORE RGB SE Wireless USB Receiver Consumer Control)
  /dev/input/event16 ignored (Corsair CORSAIR DARK CORE RGB SE Wireless USB Receiver)
  /dev/input/event17 ignored (Corsair CORSAIR DARK CORE RGB SE Wireless USB Receiver)
  /dev/input/event18 ignored (Corsair CORSAIR DARK CORE RGB SE Wireless USB Receiver)
  /dev/input/event19 grabbing failed (Corsair CORSAIR DARK CORE RGB SE Wireless USB Receiver Keyboard)
  /dev/input/event2 grabbing failed (Logitech MX Ergo)
  /dev/input/event3 ignored (Razer Razer DeathAdder V2)
  /dev/input/event4 grabbing failed (Razer Razer DeathAdder V2 Keyboard)
  /dev/input/event6 ignored (Razer Razer DeathAdder V2)
  /dev/input/event20 grabbing failed (Razer Razer DeathAdder V2)
  /dev/input/event21 ignored (HDA NVidia HDMI/DP,pcm=3)
  /dev/input/event22 ignored (HDA NVidia HDMI/DP,pcm=7)
  /dev/input/event23 ignored (HDA NVidia HDMI/DP,pcm=8)
  /dev/input/event24 ignored (HDA NVidia HDMI/DP,pcm=9)
  /dev/input/event25 ignored (HD-Audio Generic Front Mic)
  /dev/input/event26 ignored (HD-Audio Generic Rear Mic)
  /dev/input/event27 ignored (HD-Audio Generic Line)
  /dev/input/event28 ignored (HD-Audio Generic Line Out)
  /dev/input/event29 ignored (HD-Audio Generic Front Headphone)
  /dev/input/event30 already grabbed (keyd virtual keyboard)
  /dev/input/event31 ignored (keyd virtual pointer)
  /dev/input/event5 ignored (Keymapper)


This lines make me curious, becaus thats my keyboard:
  /dev/input/event9 grabbing failed (Dygma Raise Keyboard)
  /dev/input/event12 ignored (Dygma Raise)

**Reply by houmain** [reply, marked as answer]
> /dev/input/event30 grabbed (keyd virtual keyboard)

So the problem is that `keyd` is also running. This is currently not supported. Maybe the two programs are intercepting the inputs of each other.
